C/plugins/storage/sqlite/common/readings_catalogue.cpp

```diff
@@ -1960,52 +1973,73 @@ ReadingsCatalogue::tyReadingReference  ReadingsCatalogue::getReadingReference(Co
 }
 
 /**
- * Get Empty Reading Table
- *
- * @param emptyTableReference  An empty reading table reference to be used for the given asset_code
- * @return         True of success, false on any error
+ * Loads the empty reading table catalogue
  *
  */
-bool ReadingsCatalogue::getEmptyReadingTableReference(tyReadingReference &emptyTableReference)
+bool ReadingsCatalogue::loadEmptyAssetReadingCatalogue()
 {
-	bool isEmptyTableAvailable = false;
 	sqlite3 *dbHandle;
 	string sql_cmd;
 	sqlite3_stmt *stmt;
-
-	// Disable functionality temporarily to avoid regression 
-	return false;
-
 	ConnectionManager *manager = ConnectionManager::getInstance();
 	Connection *connection = manager->allocate();
 	dbHandle = connection->getDbHandle();
-	
+	m_EmptyAssetReadingCatalogue.clear();
+			
 	for (auto &item : m_AssetReadingCatalogue)
 	{
-		int tableId = item.second.first;
-		int dbId = item.second.second;
-		sql_cmd = "SELECT COUNT(*) FROM (SELECT 0 FROM readings_" + to_string(dbId) + ".readings_" + to_string(dbId) + "_" + to_string(tableId) + " LIMIT 1)";
-
+		string asset_name = item.first; // Asset
+		int tableId = item.second.first; // tableId;
+		int dbId = item.second.second; // dbId;
+		
+		sql_cmd = "SELECT COUNT(*) FROM readings_" + to_string(dbId) + ".readings_" + to_string(dbId) + "_" + to_string(tableId) + " ;";
 		if (sqlite3_prepare_v2(dbHandle, sql_cmd.c_str(), -1, &stmt, NULL) != SQLITE_OK)
 		{
-			raiseError("getEmptyReadingTableReference", sqlite3_errmsg(dbHandle));
-			return false;
+			sqlite3_finalize(stmt);	
+			continue;
 		}
 
 		if (SQLStep(stmt) == SQLITE_ROW)
 		{
 			if (sqlite3_column_int(stmt, 0) == 0)
 			{
-				isEmptyTableAvailable = true;
-				emptyTableReference.dbId = dbId;
-				emptyTableReference.tableId = tableId;
+				auto newItem = make_pair(tableId,dbId);
+				auto newMapValue = make_pair(asset_name,newItem);
+				m_EmptyAssetReadingCatalogue.insert(newMapValue);
+				
 			}
 		}
 		sqlite3_finalize(stmt);
+		
 	}
-
 	manager->release(connection);
-	return isEmptyTableAvailable;
+	return true;
+}
+
+/**
+ *  Get Empty Reading Table
+ *
+ * @param    asset emptyAsset, copies value of asset for which empty table is found
+ * @return         the reading id associated to the provided empty table
+ */
+ReadingsCatalogue::tyReadingReference ReadingsCatalogue::getEmptyReadingTableReference(std::string& asset)
+{
+	ReadingsCatalogue::tyReadingReference emptyTableReference = {-1, -1};
+	if (m_EmptyAssetReadingCatalogue.size() == 0)
+	{
+		loadEmptyAssetReadingCatalogue();
+	}
+	
+	auto it = m_EmptyAssetReadingCatalogue.begin();
+	if (it != m_EmptyAssetReadingCatalogue.end())
+	{
+		asset = it->first;
+		emptyTableReference.tableId = it->second.first;
```

Comment:
if there will be only one value in the map , and iteration over a set of values not require d, then what is the need of map data structure?

Reply:
Purge task keeps purging data on a regular interval, that's why map is being used to track of all the empty table.